File: app/src/main/java/ch/bailu/aat/gpx/xml_parser/BOM.java
package ch.bailu.aat.gpx.xml_parser;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;

import ch.bailu.foc.Foc;

public class BOM {
    public static final int BOM_UTF8    = 0xEFBBBF;
    public static final int BOM_UTF16BE = 0xFEFF;
    public static final int BOM_UTF16LE = 0xFFFE;
    public static final int BOM_UTF32BE = 0x0000FEFF;
    public static final int BOM_UTF32LE = 0xFFFE0000;


    public static boolean hasBOM(Reader reader) throws IOException {
        final int x = reader.read();

        return ((x > -1) && isBOM(x));
    }

    private static boolean isBOM(int in) {
        return  in == BOM_UTF8    ||
                in == BOM_UTF16BE ||
                in == BOM_UTF16LE ||
                in == BOM_UTF32BE ||
                in == BOM_UTF32LE;
    }


    public static Reader open(Foc file) throws IOException {
        Reader reader = new InputStreamReader(file.openR());

        if (!hasBOM(reader)) {
            reader.close();
            reader = new InputStreamReader(file.openR());

        }
        return new BufferedReader(reader);
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/file/xml/parser/BOM.kt
package ch.bailu.aat_lib.file.xml.parser

import ch.bailu.foc.Foc
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.Reader

object BOM {
    private const val BOM_UTF8 = 0xEFBBBF
    private const val BOM_UTF16LE = 0xFFFE
    private const val BOM_UTF32BE = 0x0000FEFF
    private const val BOM_UTF32LE = -0x20000

    @Throws(IOException::class)
    fun hasBOM(reader: Reader): Boolean {
        val x = reader.read()
        return x > -1 && isBOM(x)
    }

    private fun isBOM(`in`: Int): Boolean {
        return `in` == BOM_UTF8 || `in` == BOM_UTF16LE || `in` == BOM_UTF32BE || `in` == BOM_UTF32LE
    }

    @Throws(IOException::class)
    fun open(file: Foc): Reader {
        var reader: Reader = InputStreamReader(file.openR())
        if (!hasBOM(reader)) {
            reader.close()
            reader = InputStreamReader(file.openR())
        }
        return BufferedReader(reader)
    }
}
